.github/workflows/build_test.sh

```diff
@@ -108,7 +113,7 @@ for args in "${ARGS[@]}"; do
     SECONDS=0
 
     info "Checking build with $args"
-    if ! AR="$AR" CC="$CC" CXX="$CXX" CFLAGS="-Werror" CXXFLAGS="-Werror" meson -Dtests=unsafe -Dslow-tests=true -Dfuzz-tests=true --werror $args build; then
+    if ! AR="$AR" CC="$CC" CXX="$CXX" CFLAGS="-Werror" CXXFLAGS="-Werror" RUSTFLAGS="--deny warnings" meson -Dtests=unsafe -Dslow-tests=true -Dfuzz-tests=true --werror ${RUST} -Drust_args="-Clinker-plugin-lto -Copt-level=3" $args build; then
```

Comment:
I'm not sure `RUSTFLAGS` can be mixed with `-Drust_args`. As far as I know, `-Drust_args` takes precedence over `RUSTFLAGS`.

FWIW to get `-Clinker-plugin-lto -Copt-level=3` to work on Ubuntu Focal (which is used on GitHub Actions) I had to install `clang-12` and `rustc-1.52.1` and use them instead of the packages shipped by Ubuntu.

Reply:
Also, it looks like `-Clinker-plugin-lto` and `-Copt-level=3` are incompatible with `gcc` and they shouldn't be passed when `systemd` is built with `gcc`.

Reply:
Naturally, since (my understanding is) `-Clinker-plugin-lto` tells rustc to emit LLVM bitcode *instead of* the actual binary, which GCC of course doesn't know what to do with. `-Copt-level=3` should be harmless.

Reply:
thanks, updated - bash's quote escaping is being _a bit_ annoying

Reply:
> -Copt-level=3 should be harmless

I don't think it should always be passed to `rustc` though. The rust flags should probably match the C args:
```
    "--optimization=0 -Drust_args='-Copt-level=0'"
    "--optimization=2 -Drust_args='-Copt-level=2'"
    "--optimization=s -Drust_args='-Copt-level=s'"
    "--optimization=3 -Db_lto=true -Drust_args='-Copt-level=3 -Clinker-plugin-lto'"
    "--optimization=3 -Db_lto=false -Drust_args='-Copt-level=3'"
    "--optimization=3 -Ddns-over-tls=openssl -Drust_args='-Copt-level=3'"
    "--optimization=3 -Dfexecve=true -Dstandalone-binaries=true -Dstatic-libsystemd=true -Dstatic-libudev=true -Drust_args='-Copt-level=3'"
    "-Db_ndebug=true -Drust_args='-Cdebug-assertions=false'"
```
Though it would probably be better it `meson` could add Rust counterparts automatically for us in C/Rust projects.

Reply:
> Though it would probably be better it `meson` could add Rust counterparts automatically for us in C/Rust projects.

Yeah, especially given gcc-rust uses GCC-style flags.

Reply:
I tried many combinations of that, but it doesn't want to work - bash somehow always screws up the quoting, so meson fails. Tried even weird stuff like `$'...'`, but no luck.

Reply:
I think something like `eval "meson ... $args .."` instead of `meson .. $args ..` should do the trick.
```diff
diff --git a/.github/workflows/build_test.sh b/.github/workflows/build_test.sh
index 5b3eb65418..48c4e7f403 100755
--- a/.github/workflows/build_test.sh
+++ b/.github/workflows/build_test.sh
@@ -119,7 +119,7 @@ for args in "${ARGS[@]}"; do
     fi

     info "Checking build with $args"
-    if ! AR="$AR" CC="$CC" CXX="$CXX" CFLAGS="-Werror" CXXFLAGS="-Werror" RUSTFLAGS="${RUSTFLAGS}" meson -Dtests=unsafe -Dslow-tests=true -Dfuzz-tests=true --werror ${RUST} $args build; then
+    if ! AR="$AR" CC="$CC" CXX="$CXX" CFLAGS="-Werror" CXXFLAGS="-Werror" RUSTFLAGS="${RUSTFLAGS}" eval "meson -Dtests=unsafe -Dslow-tests=true -Dfuzz-tests=true --werror ${RUST} $args build"; then
         fatal "meson failed with $args"
     fi

```

Reply:
if the lto option has to be enabled only for clang, updating ARGS doesn't work either tho?

Reply:
I think it does if `ARGS` is updated depending on the compiler
```
ARGS=(
    "--optimization=0 -Drust_args='-Copt-level=0'"
    "--optimization=2 -Drust_args='-Copt-level=2'"
    "--optimization=s -Drust_args='-Copt-level=s'"
    "--optimization=3 -Db_lto=false -Drust_args='-Copt-level=3'"
    "--optimization=3 -Ddns-over-tls=openssl -Drust_args='-Copt-level=3'"
    "--optimization=3 -Dfexecve=true -Dstandalone-binaries=true -Dstatic-libsystemd=true -Dstatic-libudev=true -Drust_args='-Copt-level=3'"
    "-Db_ndebug=true -Drust_args='-Cdebug-assertions=false'"
)

if [[ "$COMPILER" == "clang" ]]; then
      ARGS+=("--optimization=3 -Db_lto=true -Drust_args='-Copt-level=3 -Clinker-plugin-lto'")
else
      ARGS+=("--optimization=3 -Db_lto=true -Drust_args='-Copt-level=3")
fi
```

Reply:
Then again, all those kludges are necessary just because `meson` doesn't seem to support C/Rust projects. I think ideally it should probably be fixed there eventually.

Reply:
ok, let's see if that works

Reply:
Looks like `-Cdebug-assertions=false` isn't exactly what `rustc` expects:
```
FAILED: src/basic/libbasic_rust.a 
rustc -C linker=clang-11 --color=always --crate-type staticlib --crate-name basic_rust -g --emit dep-info=src/basic/basic_rust.d --emit link -o src/basic/libbasic_rust.a --deny warnings -Cdebug-assertions=false ../src/basic/string-util.rs
error: incorrect value `false` for codegen option `debug-assertions` - one of: `y`, `yes`, `on`, `n`, `no`, or `off` was expected
```

On the bright side though it seems the "eval" trick is working and all the arguments are passed more or less correctly :-)

Reply:
yep, thanks for the suggestion! I did not think about `eval` at all

Reply:
looks like it's working!